clients/drcachesim/tests/cache_lru_unit_tests.cpp

```diff
@@ -0,0 +1,106 @@
+ *   and/or other materials provided with the distribution.
+ *
+ * * Neither the name of Google, Inc. nor the names of its contributors may be
+ *   used to endorse or promote products derived from this software without
+ *   specific prior written permission.
+ *
+ * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+ * AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+ * IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+ * ARE DISCLAIMED. IN NO EVENT SHALL GOOGLE, INC. OR CONTRIBUTORS BE LIABLE
+ * FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+ * DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+ * SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ * CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ * LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+ * OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+ * DAMAGE.
+ */
+
+// Unit tests for cache_lru
+#include <iostream>
+#include <assert.h>
+#include "simulator/cache_lru.h"
+
+struct cache_config_t {
+    int associativity = 4;
+    int line_size = 32;
+    int total_size = 256;
+    caching_device_t *parent = nullptr;
+    caching_device_stats_t *stats = new cache_stats_t(line_size, "", true);
+    prefetcher_t *prefetcher = nullptr;
+    bool inclusive = true;
+    bool coherent_cache = true;
+    int id = -1;
+    snoop_filter_t *snoop_filter = nullptr;
+};
+
+static cache_config_t
+make_test_configs(int associativity, int line_size, int total_size)
+{
+    cache_config_t config;
+    config.associativity = associativity;
+    config.line_size = line_size;
+    config.total_size = total_size;
+    return config;
+}
+
+class cache_lru_test_t : public cache_lru_t {
+public:
+    void
+    unit_test_access_update()
+    {
+        // Create and initialize an 8-way set associative cache with line size of 32 and
+        // total size of 1024 bytes.
+        cache_config_t config_two = make_test_configs(8, 32, 1024);
+        if (!init(config_two.associativity, config_two.line_size, config_two.total_size,
+                  config_two.parent, config_two.stats, config_two.prefetcher,
+                  config_two.inclusive, config_two.coherent_cache, config_two.id,
+                  config_two.snoop_filter)) {
+            std::cerr << "LRU cache failed to initialize"
+                      << "\n";
+            exit(1);
+        }
+
+        // Access the cache lines in set 1 in the following sequence and ensure the
+        // counters are updated properly.
+        access_update(1, 6);
+        access_update(1, 1);
+        access_update(1, 2);
+        access_update(1, 3);
+        access_update(1, 4);
+        access_update(1, 5);
+        access_update(1, 0);
+        access_update(1, 7);
+
+        assert(get_caching_device_block(1, 0).counter_ == 1);
```

Comment:
I think it would be more robust and maintainable to have a black-box test that doesn't know/care about *how* LRU is implemented with this private counter: it would just makes sure the least-recently-used line is replaced.  So it would ask replace_which_way() and ensure it got the right answer, rather than making assumptions about how LRU is implemented using this internal counter variable.

Then we ensure that what matters, replace_which_way(), is operating how we want it to: which this test here doesn't actually test, since we could have perfect counter udpate code but a bug in replace_which_way and this wouldn't find it.

Plus, we could then swap to a different internal counter algorithm for LRU and have the same test, as it would focus on the interface: the contract this LRU module is providing.

Reply:
When testing out this change with CL https://github.com/DynamoRIO/dynamorio/pull/4883 reverted I realized `replace_which_way()` could pass (returns the `way` with highest counter) even though the counters are not set properly in the first place. For example consider the 8-way set associative cache used in this test. With https://github.com/DynamoRIO/dynamorio/pull/4883 reverted, when accessing the ways sequentially 0 through 7, the counter values will be as follows:

| way | counter |
| ----|-------- |
|  0  |    2    |
|  1  |    2    |
|  2  |    2    |
|  3  |    2    |
|  4  |    2    |
|  5  |    2    |
|  6  |    1    |
|  7  |    0    |

As opposed to the following correct ones:
| way | counter |
| ----|-------- |
|  0  |    7    |
|  1  |    6    |
|  2  |    5    |
|  3  |    4    |
|  4  |    3    |
|  5  |    2    |
|  6  |    1    |
|  7  |    0    |

That is obviously wrong but `replace_which_way()` would still return the correct way to be replaced (0). The reason for that is even though the actual counter is wrong, it is still the highest count. This masks the bug fixed by https://github.com/DynamoRIO/dynamorio/pull/4883 fixed so decided to keep the counter test in there as well which this unit test would catch as expected.

Reply:
I would expect to test the actual sequence from #4881 that exposes the bug?  Isn't the problem that the test in this PR only tests one simple sequence and it's not a sequence that actually hits the bug?